zowe/zowe-install-packaging
# zwe ts edition, with configmgr

This PR introduces the configmgr into the zowe build, and uses it in an all-new port of the existing zwe scripts over to typescript which the configmgr can execute.
The ts code (running as js after build) acts nearly identical to the existing shell code, and it should be possible to get 100% compatibility in the near future. 

**The typescript code is not running under node. configmgr has its own js interpreter so there is no prereq or dependency**

configmgr has advantages by itself, such as validating configuration against schemas, templating within the config file, and allowing for multiple configuration files in a tiered list that gets merged into 1 unified config for server use, so perhaps we could put our defaults in one file and overrides in another.

The ts code builds upon these advantages to also enable compile-time checks so we can spot & fix bugs more easily, but also it's much, much faster than the shell code at doing equivalent actions.

So far, the ts code has only been written to implement "zwe internal start prepare". No other phases are implemented yet but can be over time.

**This code has no impact unless you opt-in to it.** If you are just using `zwe` the same way you always did, you will get the same behavior as before.
If you instead set the env var `ZWE_CLI_CONFIGMGR=true`, then `zwe` commands will use the ts+configmgr code paths, at least for the `zwe` commands we've implemented so far. The rest will just run the usual shell scripts.

**This has no change to components with/without `ZWE_CLI_CONFIGMGR=true` in that they get the same execution environment, and barring bugs, same workspace contents and same env vars.**

Each ts file exists in the same directory as the shell counterparts.
In the `bin/commands` area, each `zwe` command starts as a shell script and can switch over to TS depending on the env var. When it does this, at the moment a `cli.ts` is used to bridge the gap between shell code and ts code.


**This PR depends on various small PRs from other repos, see https://github.com/zowe/zowe-install-packaging/issues/2662 for the list**

**A recording showing how this code works and what benefits it as is here https://github.com/zowe/zowe-install-packaging/issues/2879**

## Testing
Try running `zwe start`, you should see no difference in behavior.
Now get a clean workspace directory and run  `zwe internal start prepare`
Then compare it to the workspace made by `ZWE_CLI_CONFIGMGR=true zwe internal start prepare`
Additionally, try the new features:

- Make a typo in zowe.yaml, like a bad port or a typo in the keyring name type PKCS12, or missing a required property.  The "prepare" should fail with a warning saying verification failed and pointing to what in particular the issue was.
- Try templating in zowe.yaml (format `${{ js-goes-here }}`). Try for example, mapping the externalPort to the gateway port, such as by setting the externalPort to ` ${{ components.gateway.port }}`
- Try multiple config files. instead of `zwe internal start prepare -c zowe.yaml` do `ZWE_CLI_CONFIGMGR=true zwe internal start prepare -c FILE(override.yaml):FILE(base.yaml)`

If you'd like to inspect the result of templating and merging, look for the file `workspaceDirectory/.env/zowe-merged.yaml`.
Also try looking at the envs in `workspaceDirectory/.env` in general. **If you spot a difference, let us know!**

**note: you can't go between ZWE_CLI_CONFIGMGR=true and ZWE_CLI_CONFIGMGR=false once you start using templating and multi-file, because the shell scripts dont support it. You can swap between these modes if you use just an ordinary zowe.yaml though**





- [ ] Necessary documentation (if appropriate) have been added / updated
- [x] DCO signoffs have been added to all commits, including this PR

#### PR type

#### Relevant issues
See https://github.com/zowe/zowe-install-packaging/issues/2662
And https://github.com/zowe/zowe-install-packaging/issues/2879

#### Changes proposed in this PR
- Bundles the 'configmgr' binary for zos. Configmgr should compile on other platforms but we'll get there soon.
- Updates schema files upon testing them
- Uses configmgr to read the schema files to validate zowe.yaml and manifest.yaml files
- Uses configmgr's builtin JS engine to run this logic and overall all logic of "zwe internal start prepare" within the context of TS scripts. This PR contains a TS equivalent of most shell scripts within /bin/libs and the behavior is nearly identical (just a few TODOs in code) and can be 100% identical in the near future.
- This PR does not apply the TS code to every single zwe command, but it can be done over time.
- The TS code and configmgr do not execute unless the env var "ZWE_CLI_CONFIGMGR=true". Although the configmgr should be backward compatible, in case of bugs we can just disable it. For now, we can consider it opt-in feature until we get more coverage & confidence in it.

#### Does this PR introduce a breaking change?

#### Does this PR do something the person installing Zowe should know about?
I will be updating the README of the bin folder to explain how this code works, soon.
